x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticationToken.java

```diff
@@ -16,35 +16,35 @@
 import java.text.ParseException;
 import java.util.Arrays;
 import java.util.Objects;
+import java.util.TreeSet;
 
 /**
  * An {@link AuthenticationToken} to hold JWT authentication related content.
  */
 public class JwtAuthenticationToken implements AuthenticationToken {
-    private final String principal;
+    private static final char PRINCIPAL_SEPARATOR = '|';
+    private static final char PRINCIPAL_LITERAL_SEPARATOR_REPLACE = ' ';
```

Comment:
I've chosen to use `|` as a separator between the claims that form the principal, instead of `/`, because the claims can contain URLS that have a lot of `/`s.
I've also chosen to replace any existing `|` in the claims themselves with ` ` to remove any ambiguity when parsing the JWT principal (e.g. from the audit logs).

Reply:
Good point about the tokenization. I've followed your suggestion.
It only uses the `String` or `String[]` claim values, the rest are ignored.
If the claim value is a `String[]` then the values are separated by `,`.
I've not escaped or remove any exiting `:`, `'` or ` ` in the claim name or value.